File: samlang-interpreter/src/commonMain/kotlin/samlang/interpreter/MidIrInterpreter.kt
package samlang.interpreter

import samlang.ast.common.IrNameEncoder
import samlang.ast.common.IrOperator
import samlang.ast.mir.MidIrCompilationUnit
import samlang.ast.mir.MidIrExpression
import samlang.ast.mir.MidIrFunction
import samlang.ast.mir.MidIrLoweredExpressionVisitor
import samlang.ast.mir.MidIrLoweredStatementVisitor
import samlang.ast.mir.MidIrStatement

/**
 * Interpret [compilationUnit] using [entryModule]'s main function as the main function.
 *
 * @return the printed string.
 */
fun interpretCompilationUnit(compilationUnit: MidIrCompilationUnit): String {
    val environment = setupEnvironment(compilationUnit = compilationUnit)
    val function = environment.functions[IrNameEncoder.compiledProgramMain] ?: error(message = "Bad function.")
    interpretFunction(irFunction = function, environment = environment, arguments = emptyList())
    return environment.printed.toString()
}

private fun setupEnvironment(compilationUnit: MidIrCompilationUnit): GlobalEnvironment {
    val functions = compilationUnit.functions.map { it.functionName to it }.toMap()
    val globalVariables = mutableMapOf<String, Long>()
    val strings = mutableMapOf<Long, String>()
    var heapPointer = 10000L
    compilationUnit.globalVariables.forEach { (name, content) ->
        val location = heapPointer
        globalVariables[name] = location
        strings[location + 8] = content
        heapPointer += 8
    }
    val functionsGlobals = mutableMapOf<Long, String>()
    compilationUnit.functions.forEach { function ->
        val location = heapPointer
        val name = function.functionName
        globalVariables[name] = location
        functionsGlobals[location] = name
        heapPointer += 8
    }
    return GlobalEnvironment(
        functions = functions,
        globalVariables = globalVariables,
        strings = strings,
        functionsGlobals = functionsGlobals,
        heap = mutableMapOf(),
        heapPointer = heapPointer,
        printed = StringBuilder()
    )
}

private fun interpretFunction(irFunction: MidIrFunction, arguments: List<Long>, environment: GlobalEnvironment): Long {
    require(value = irFunction.argumentTemps.size == arguments.size)
    val stackFrame = StackFrame()
    irFunction.argumentTemps.map { it.id }.zip(other = arguments).forEach { (id, value) ->
        stackFrame[id] = value
    }
    val statements = irFunction.mainBodyStatements
    val interpreter = MidIrStatementInterpreter(
        environment = environment,
        expressionVisitor = MidIrExpressionInterpreter(environment = environment),
        statements = statements
    )
    var returnedValue: Long? = null
    while (returnedValue == null) {
        val statement = statements[interpreter.programCounter]
        statement.accept(visitor = interpreter, context = stackFrame)
        returnedValue = stackFrame.returnValue
    }
    return returnedValue
}

private class MidIrStatementInterpreter(
    private val environment: GlobalEnvironment,
    private val expressionVisitor: MidIrExpressionInterpreter,
    statements: List<MidIrStatement>
) : MidIrLoweredStatementVisitor<StackFrame, Unit> {
    private var _programCounter: Int = 0
    private val labelMapping: Map<String, Int>

    val programCounter: Int get() = _programCounter

    init {
        labelMapping = mutableMapOf<String, Int>().apply {
            statements.forEachIndexed { index, statement ->
                if (statement is MidIrStatement.Label) {
                    this[statement.name] = index
                }
            }
        }
    }

    override fun visit(node: MidIrStatement.MoveTemp, context: StackFrame) {
        context[node.tempId] = node.source.accept(visitor = expressionVisitor, context = context)
        _programCounter++
    }

    override fun visit(node: MidIrStatement.MoveMem, context: StackFrame) {
        environment.heap[node.memLocation.accept(visitor = expressionVisitor, context = context)] =
            node.source.accept(visitor = expressionVisitor, context = context)
        _programCounter++
    }

    override fun visit(node: MidIrStatement.CallFunction, context: StackFrame) {
        val arguments = node.arguments.map { it.accept(visitor = expressionVisitor, context = context) }
        val functionExpression = node.functionExpr
        val functionName = if (functionExpression is MidIrExpression.Name) {
            val functionName = functionExpression.name
            val result = when (functionName) {
                IrNameEncoder.nameOfMalloc -> {
                    require(value = arguments.size == 1)
                    val size = arguments[0]
                    val start = environment.heapPointer
                    environment.heapPointer += size
                    start
                }
                IrNameEncoder.nameOfThrow -> {
                    require(value = arguments.size == 1)
                    val argument = arguments[0]
                    val string = environment.strings[argument] ?: error(message = "Bad string at $argument")
                    throw PanicException(reason = string)
                }
                IrNameEncoder.nameOfStringToInt -> {
                    require(value = arguments.size == 1)
                    val argument = arguments[0]
                    val string = environment.strings[argument] ?: error(message = "Bad string at $argument")
                    string.toLongOrNull() ?: throw PanicException(reason = "Bad string: $string")
                }
                IrNameEncoder.nameOfIntToString -> {
                    require(value = arguments.size == 1)
                    val stringForm = arguments[0].toString()
                    val location = environment.heapPointer
                    environment.heapPointer += 8
                    environment.strings[location] = stringForm
                    location
                }
                IrNameEncoder.nameOfStringConcat -> {
                    require(value = arguments.size == 2)
                    val string1 = environment.strings[arguments[0]] ?: error(message = "Bad string at ${arguments[0]}")
                    val string2 = environment.strings[arguments[1]] ?: error(message = "Bad string at ${arguments[1]}")
                    val location = environment.heapPointer
                    environment.heapPointer += 8
                    environment.strings[location] = string1 + string2
                    location
                }
                IrNameEncoder.nameOfPrintln -> {
                    require(value = arguments.size == 1)
                    val argument = arguments[0]
                    val string = environment.strings[argument] ?: error(message = "Bad string at $argument")
                    environment.printed.append(string).append('\n')
                    0L
                }
                else -> null
            }
            if (result != null) {
                node.returnCollector?.let { context[it.id] = result }
                _programCounter++
                return
            }
            functionName
        } else {
            val functionAddress = functionExpression.accept(visitor = expressionVisitor, context = context)
            environment.functionsGlobals[functionAddress]
                ?: error(message = "Undefined function at $functionAddress! Expression: $functionExpression")
        }
        val function = environment.functions[functionName]
            ?: error(message = "Missing function $functionName")
        val result = interpretFunction(
            irFunction = function,
            arguments = arguments,
            environment = environment
        )
        node.returnCollector?.let { context[it.id] = result }
        _programCounter++
    }

    override fun visit(node: MidIrStatement.Jump, context: StackFrame) {
        val label = node.label
        _programCounter = (labelMapping[label] ?: error("BAD label: $label!"))
    }

    override fun visit(node: MidIrStatement.ConditionalJumpFallThrough, context: StackFrame) {
        if (node.condition.accept(visitor = expressionVisitor, context = context) != 0L) {
            visit(node = MidIrStatement.Jump(label = node.label1), context = context)
        } else {
            _programCounter++
        }
    }

    override fun visit(node: MidIrStatement.Label, context: StackFrame) {
        _programCounter++
    }

    override fun visit(node: MidIrStatement.Return, context: StackFrame) {
        val value = node.returnedExpression?.accept(visitor = expressionVisitor, context = context) ?: 0L
        context.returnValue(value = value)
    }
}

private class MidIrExpressionInterpreter(
    private val environment: GlobalEnvironment
) : MidIrLoweredExpressionVisitor<StackFrame, Long> {
    override fun visit(node: MidIrExpression.Constant, context: StackFrame): Long = node.value

    override fun visit(node: MidIrExpression.Name, context: StackFrame): Long =
        environment.globalVariables[node.name] ?: error(message = "Referencing undefined global ${node.name}!")

    override fun visit(node: MidIrExpression.Temporary, context: StackFrame): Long = context[node.id]

    override fun visit(node: MidIrExpression.Op, context: StackFrame): Long {
        val value1 = node.e1.accept(visitor = this, context = context)
        val value2 = node.e2.accept(visitor = this, context = context)
        return when (node.operator) {
            IrOperator.MUL -> value1 * value2
            IrOperator.DIV -> {
                if (value2 == 0L) {
                    throw PanicException(reason = "Division by zero!")
                }
                value1 / value2
            }
            IrOperator.MOD -> {
                if (value2 == 0L) {
                    throw PanicException(reason = "Division by zero!")
                }
                value1 % value2
            }
            IrOperator.ADD -> value1 + value2
            IrOperator.SUB -> value1 - value2
            IrOperator.XOR -> value1 xor value2
            IrOperator.LT -> toInt(boolean = value1 < value2)
            IrOperator.LE -> toInt(boolean = value1 <= value2)
            IrOperator.GT -> toInt(boolean = value1 > value2)
            IrOperator.GE -> toInt(boolean = value1 >= value2)
            IrOperator.EQ -> toInt(boolean = value1 == value2)
            IrOperator.NE -> toInt(boolean = value1 != value2)
            IrOperator.AND -> if (value1 == 0L) 0L else if (value2 == 0L) 0L else 1L
            IrOperator.OR -> if (value1 != 0L) 1L else if (value2 != 0L) 1L else 0L
        }
    }

    private fun toInt(boolean: Boolean): Long = if (boolean) 1L else 0L

    override fun visit(node: MidIrExpression.Mem, context: StackFrame): Long =
        environment.heap[node.expression.accept(visitor = this, context = context)] ?: 0
}

private class GlobalEnvironment(
    val functions: Map<String, MidIrFunction>,
    val globalVariables: Map<String, Long>,
    val functionsGlobals: Map<Long, String>,
    val strings: MutableMap<Long, String>,
    val heap: MutableMap<Long, Long>,
    var heapPointer: Long,
    val printed: StringBuilder
)

private class StackFrame {
    private val variables: MutableMap<String, Long> = mutableMapOf()
    private var _returnedValue: Long? = null

    val returnValue: Long? get() = _returnedValue

    operator fun get(name: String): Long = variables[name] ?: 0

    operator fun set(name: String, value: Long) {
        variables[name] = value
    }

    fun returnValue(value: Long) {
        _returnedValue = value
    }
}

File: samlang-core-interpreter/mid-ir-interpreter.ts
/* eslint-disable no-param-reassign */

import PanicException from './panic-exception';

import {
  ENCODED_FUNCTION_NAME_MALLOC,
  ENCODED_FUNCTION_NAME_THROW,
  ENCODED_FUNCTION_NAME_STRING_TO_INT,
  ENCODED_FUNCTION_NAME_INT_TO_STRING,
  ENCODED_FUNCTION_NAME_STRING_CONCAT,
  ENCODED_FUNCTION_NAME_PRINTLN,
  ENCODED_COMPILED_PROGRAM_MAIN,
} from 'samlang-core-ast/common-names';
import type { IROperator } from 'samlang-core-ast/common-operators';
import type {
  MidIRFunction,
  MidIRCompilationUnit,
  MidIRExpression,
} from 'samlang-core-ast/mir-nodes';
import { Long, assertNotNull, checkNotNull } from 'samlang-core-utils';

export class StackFrame {
  private variables = new Map<string, Long>();

  private _returnValue: Long | null = null;

  get returnValue(): Long | null {
    return this._returnValue;
  }

  setReturnValue(value: Long): void {
    this._returnValue = value;
  }

  getLocalValue(name: string): Long {
    return this.variables.get(name) ?? Long.ZERO;
  }

  setLocalValue(name: string, value: Long): void {
    this.variables.set(name, value);
  }
}

export type GeneralIREnvironment = {
  // Global variable name to fake address mapping.
  readonly globalVariables: ReadonlyMap<string, Long>;
  // Fake function address to function name mapping.
  readonly functionsGlobals: ReadonlyMap<string, string>;
  // Strings generated at compile time and runtime.
  readonly strings: Map<string, string>;
  // Address to value mapping of heap.
  readonly heap: Map<string, Long>;
  heapPointer: Long;
  // A collection of already printed stuff.
  printed: string;
};

type MidIRInterpreterMutableGlobalEnvironment = {
  // A collection of all available functions.
  readonly functions: ReadonlyMap<string, MidIRFunction>;
} & GeneralIREnvironment;

export const handleBuiltInFunctionCall = (
  environment: GeneralIREnvironment,
  functionName: string,
  functionArgumentValues: readonly Long[]
): Long | null => {
  switch (functionName) {
    case ENCODED_FUNCTION_NAME_MALLOC: {
      const start = environment.heapPointer;
      environment.heapPointer = environment.heapPointer.add(
        checkNotNull(functionArgumentValues[0])
      );
      return start;
    }
    case ENCODED_FUNCTION_NAME_THROW: {
      const string = environment.strings.get(checkNotNull(functionArgumentValues[0]).toString());
      // istanbul ignore next
      if (string == null) throw new Error('Bad string!');
      throw new PanicException(string);
    }
    case ENCODED_FUNCTION_NAME_STRING_TO_INT: {
      const string = environment.strings.get(checkNotNull(functionArgumentValues[0]).toString());
      // istanbul ignore next
      if (string == null) throw new Error('Bad string!');
      try {
        BigInt(string);
        return Long.fromString(string);
      } catch {
        throw new PanicException(`Bad string: ${string}`);
      }
    }
    case ENCODED_FUNCTION_NAME_INT_TO_STRING: {
      const string = String(functionArgumentValues[0]);
      const location = environment.heapPointer;
      environment.heapPointer = environment.heapPointer.add(8);
      environment.strings.set(location.toString(), string);
      return location;
    }
    case ENCODED_FUNCTION_NAME_STRING_CONCAT: {
      const string1 = environment.strings.get(checkNotNull(functionArgumentValues[0]).toString());
      const string2 = environment.strings.get(checkNotNull(functionArgumentValues[1]).toString());
      // istanbul ignore next
      if (string1 == null || string2 == null) throw new Error('Bad string');
      const location = environment.heapPointer;
      environment.heapPointer = environment.heapPointer.add(8);
      environment.strings.set(location.toString(), string1 + string2);
      return location;
    }
    case ENCODED_FUNCTION_NAME_PRINTLN: {
      const string = environment.strings.get(checkNotNull(functionArgumentValues[0]).toString());
      // istanbul ignore next
      if (string == null) throw new Error('Bad string!');
      environment.printed += `${string}\n`;
      return Long.ZERO;
    }
    default:
      return null;
  }
};

const longOfBool = (b: boolean) => (b ? Long.ONE : Long.ZERO);

export const computeBinary = (operator: IROperator, value1: Long, value2: Long): Long => {
  switch (operator) {
    case '+':
      return value1.add(value2);
    case '-':
      return value1.subtract(value2);
    case '*':
      return value1.multiply(value2);
    case '/':
      if (value2.equals(Long.ZERO)) throw new PanicException('Division by zero!');
      return value1.divide(value2);
    case '%':
      if (value2.equals(Long.ZERO)) throw new PanicException('Mod by zero!');
      return value1.mod(value2);
    case '^':
      return value1.xor(value2);
    case '<':
      return longOfBool(value1.lessThan(value2));
    case '<=':
      return longOfBool(value1.lessThanOrEqual(value2));
    case '>':
      return longOfBool(value1.greaterThan(value2));
    case '>=':
      return longOfBool(value1.greaterThanOrEqual(value2));
    case '==':
      return longOfBool(value1.equals(value2));
    case '!=':
      return longOfBool(value1.notEquals(value2));
  }
};

const interpretMidIRExpression = (
  environment: MidIRInterpreterMutableGlobalEnvironment,
  stackFrame: StackFrame,
  expression: MidIRExpression
): Long => {
  switch (expression.__type__) {
    case 'MidIRConstantExpression':
      return expression.value;
    case 'MidIRNameExpression': {
      const value = environment.globalVariables.get(expression.name);
      // istanbul ignore next
      if (value == null) throw new Error(`Referencing undefined global ${expression.name}.`);
      return value;
    }
    case 'MidIRTemporaryExpression':
      return stackFrame.getLocalValue(expression.name);
    case 'MidIRImmutableMemoryExpression': {
      const value = environment.heap.get(
        interpretMidIRExpression(environment, stackFrame, expression.indexExpression).toString()
      );
      // istanbul ignore next
      if (value == null) throw new Error();
      return value;
    }
    case 'MidIRBinaryExpression': {
      const value1 = interpretMidIRExpression(environment, stackFrame, expression.e1);
      const value2 = interpretMidIRExpression(environment, stackFrame, expression.e2);
      return computeBinary(expression.operator, value1, value2);
    }
  }
};

const interpretMidIRFunction = (
  environment: MidIRInterpreterMutableGlobalEnvironment,
  midIRFunction: MidIRFunction,
  functionArguments: readonly Long[]
): Long => {
  // istanbul ignore next
  if (functionArguments.length !== midIRFunction.argumentNames.length) throw new Error();
  const stackFrame = new StackFrame();
  midIRFunction.argumentNames.forEach((argumentName, index) => {
    stackFrame.setLocalValue(argumentName, checkNotNull(functionArguments[index]));
  });

  let programCounter = 0;
  const labelMapping = new Map<string, number>();
  midIRFunction.mainBodyStatements.forEach((statement, index) => {
    if (statement.__type__ === 'MidIRLabelStatement') {
      labelMapping.set(statement.name, index);
    }
  });

  let returnedValue: Long | null = null;
  while (returnedValue == null) {
    const statementToInterpret = midIRFunction.mainBodyStatements[programCounter];
    assertNotNull(statementToInterpret);

    switch (statementToInterpret.__type__) {
      case 'MidIRMoveTempStatement':
        stackFrame.setLocalValue(
          statementToInterpret.temporaryID,
          interpretMidIRExpression(environment, stackFrame, statementToInterpret.source)
        );
        programCounter += 1;
        break;

      case 'MidIRMoveMemStatement':
        environment.heap.set(
          interpretMidIRExpression(
            environment,
            stackFrame,
            statementToInterpret.memoryIndexExpression
          ).toString(),
          interpretMidIRExpression(environment, stackFrame, statementToInterpret.source)
        );
        programCounter += 1;
        break;

      case 'MidIRJumpStatement': {
        const target = labelMapping.get(statementToInterpret.label);
        // istanbul ignore next
        if (target == null) throw new Error(`Bad label ${statementToInterpret.label}!`);
        programCounter = target;
        break;
      }

      case 'MidIRLabelStatement':
        programCounter += 1;
        break;

      case 'MidIRConditionalJumpFallThrough':
        if (
          interpretMidIRExpression(
            environment,
            stackFrame,
            statementToInterpret.conditionExpression
          ).notEquals(Long.ZERO)
        ) {
          const target = labelMapping.get(statementToInterpret.label1);
          // istanbul ignore next
          if (target == null) throw new Error(`Bad label ${statementToInterpret.label1}!`);
          programCounter = target;
        } else {
          programCounter += 1;
        }
        break;

      case 'MidIRReturnStatement':
        stackFrame.setReturnValue(
          interpretMidIRExpression(environment, stackFrame, statementToInterpret.returnedExpression)
        );
        break;

      case 'MidIRCallFunctionStatement': {
        const functionArgumentValues = statementToInterpret.functionArguments.map((it) =>
          interpretMidIRExpression(environment, stackFrame, it)
        );
        const functionExpression = statementToInterpret.functionExpression;
        let functionName: string;
        if (functionExpression.__type__ === 'MidIRNameExpression') {
          functionName = functionExpression.name;
          const result = handleBuiltInFunctionCall(
            environment,
            functionName,
            functionArgumentValues
          );
          if (result != null) {
            if (statementToInterpret.returnCollectorTemporaryID != null) {
              stackFrame.setLocalValue(statementToInterpret.returnCollectorTemporaryID, result);
            }
            programCounter += 1;
            break;
          }
        } else {
          const functionAddress = interpretMidIRExpression(
            environment,
            stackFrame,
            functionExpression
          );
          const nullableName = environment.functionsGlobals.get(functionAddress.toString());
          // istanbul ignore next
          if (nullableName == null) throw new Error(`Undefined function at ${functionAddress}!`);
          functionName = nullableName;
        }

        const functionToCall = environment.functions.get(functionName);
        // istanbul ignore next
        if (functionToCall == null) throw new Error(`Missing function ${functionName}`);
        const result = interpretMidIRFunction(environment, functionToCall, functionArgumentValues);
        if (statementToInterpret.returnCollectorTemporaryID != null) {
          stackFrame.setLocalValue(statementToInterpret.returnCollectorTemporaryID, result);
        }
        programCounter += 1;
        break;
      }
    }

    returnedValue = stackFrame.returnValue;
  }

  return returnedValue;
};

const setupMidIRCompilationUnitIntepretationEnvironment = (
  compilationUnit: MidIRCompilationUnit
): MidIRInterpreterMutableGlobalEnvironment => {
  const functions = new Map(compilationUnit.functions.map((it) => [it.functionName, it]));
  const globalVariables = new Map<string, Long>();
  const strings = new Map<string, string>();
  let heapPointer = Long.fromInt(10000);
  compilationUnit.globalVariables.forEach(({ name, content }) => {
    const location = heapPointer;
    globalVariables.set(name, location);
    strings.set(location.toString(), content);
    heapPointer = heapPointer.add(Long.fromInt(8));
  });
  const functionsGlobals = new Map<string, string>();
  compilationUnit.functions.forEach(({ functionName }) => {
    const location = heapPointer;
    globalVariables.set(functionName, location);
    functionsGlobals.set(location.toString(), functionName);
    heapPointer = heapPointer.add(Long.fromInt(8));
  });
  return {
    functions,
    globalVariables,
    strings,
    functionsGlobals,
    heap: new Map(),
    heapPointer,
    printed: '',
  };
};

const interpretMidIRCompilationUnit = (compilationUnit: MidIRCompilationUnit): string => {
  const environment = setupMidIRCompilationUnitIntepretationEnvironment(compilationUnit);
  const mainFunction = environment.functions.get(ENCODED_COMPILED_PROGRAM_MAIN);
  // istanbul ignore next
  if (mainFunction == null) throw new Error('Missing new function!');
  interpretMidIRFunction(environment, mainFunction, []);
  return environment.printed;
};

export default interpretMidIRCompilationUnit;
